server/apps/microservices/src/processors/video-transcode.processor.ts

```diff
@@ -72,10 +73,45 @@ export class VideoTranscodeProcessor {
   async runVideoEncode(asset: AssetEntity, savedEncodedPath: string): Promise<void> {
     const config = await this.systemConfigService.getConfig();
 
-    if (config.ffmpeg.transcode == TranscodePreset.ALL) {
+    const transcode = await this.needsTranscoding(asset, config.ffmpeg);
+    if (transcode) {
+      //TODO: If video or audio are already the correct format, don't re-encode, copy the stream
       return this.runFFMPEGPipeLine(asset, savedEncodedPath);
     }
+  }
+
+  async needsTranscoding(asset: AssetEntity, ffmpegConfig: SystemConfigFFmpegDto): Promise<boolean> {
+    switch (ffmpegConfig.transcode) {
+      case TranscodePreset.ALL:
+        return true;
+
+      case TranscodePreset.REQUIRED:
+        {
+          const videoStream = await this.getVideoStream(asset);
+          if (videoStream.codec_name !== ffmpegConfig.targetVideoCodec) {
+            return true;
+          }
+        }
+        break;
+
+      case TranscodePreset.OPTIMAL:
+        {
+          const videoStream = await this.getVideoStream(asset);
+          if (videoStream.codec_name !== ffmpegConfig.targetVideoCodec) {
+            return true;
+          }
+
+          const videoHeightThreshold = 1080;
+          if (!videoStream.height || videoStream.height > videoHeightThreshold) {
+            return true;
+          }
```

Comment:
This can just be `return !videoStream.height || videoStream.height > videoHeightThreshold`

Reply:
Oh, sure. You're right

Reply:
Done!